pixeltable/exec/exec_context.py

```diff
@@ -1,42 +1,210 @@
+import builtins
+import logging
 import random
+import time
 
 import sqlalchemy as sql
+from rich.progress import Column, Progress, ProgressColumn, Task, TaskID, Text, TextColumn
 
 from pixeltable import exprs
+from pixeltable.env import Env
+
+_logger = logging.getLogger('pixeltable')
 
 
 class ExecContext:
     """Class for execution runtime constants"""
 
     row_builder: exprs.RowBuilder
+    show_progress: bool
+    progress: Progress | None
+    progress_start: float  # time.monotonic() of progress.start()
+    progress_reporters: dict[str, 'ExecContext.ProgressReporter']
+    elapsed_time_task_id: TaskID | None
+    batch_size: int  # 0: no batching
     profile: exprs.ExecProfile
-    show_pbar: bool
-    batch_size: int
-    num_rows: int | None
-    conn: sql.engine.Connection | None
+    conn: sql.engine.Connection | None  # if present, use this to execute SQL queries
     pk_clause: list[sql.ClauseElement] | None
-    num_computed_exprs: int
     ignore_errors: bool
     random_seed: int  # general-purpose source of randomness with execution scope
 
+    class ProgressReporter:
+        """
+        Represents a single Task, attached to ExecCtx.progress.
+
+        Task creation is deferred until the first update() call, in order to avoid useless output.
+        """
+
+        task_id: TaskID | None
+        ctx: 'ExecContext'
+        last_update_ts: float | None
+        reports_bytes: bool  # if True, automatically scales the reported numbers to human-readable units
+        total: float
+        desc: str
+        unit: str
+
+        def __init__(self, ctx: 'ExecContext', desc: str, unit: str):
+            self.ctx = ctx
+            self.desc = desc
+            self.unit = unit
+            self.reports_bytes = unit == 'B'
+            self.task_id = None
+            self.last_update_ts = None
+            self.total = 0
+
+        def _create_task(self) -> None:
+            if self.task_id is None:
+                self.task_id = self.ctx.progress.add_task(self.desc, rate='0/s', unit=self.unit)
+                self.last_update_ts = time.monotonic()  # start now
+
+        def _get_display_unit(self) -> tuple[int, str]:
+            # scale to human-readable unit
+            scale: int
+            unit: str
+            if self.total < 2**10:
+                scale = 0
+                unit = 'B'
```

Comment:
Suggestion: use `'B '` (with a space at the end) for consistent alignment

Reply:
Done.